Fix "Undefined array key" warnings in modPhpThumb

jolichter opened the pull request:
Fix PHP warning: "undefined array key ..."  HTTP_IF_MODIFIED_SINCE, sia and down

### What does it do?
Prevents 3 modPhpThumb.php warnings when uploading files via the media browser

### Why is it needed?
no PHP warnings in MODX error log

### How to test
Describe how to test the changes you made.

### Related issue(s)/PR(s)
Please also refer to [16340](https://github.com/modxcms/revolution/issues/16340)

jolichter committed:
Update modPhpThumb.php

Fix PHP warning: "undefined array key ..."  HTTP_IF_MODIFIED_SINCE, sia and down

jolichter commented:
 

 
 

Gesendet: Montag, 26. Dezember 2022 um 20:22 Uhr
Von: "cla-bot[bot]" ***@***.***>
An: "modxcms/revolution" ***@***.***>
Cc: "Jo Lichter" ***@***.***>, "Mention" ***@***.***>
Betreff: Re: [modxcms/revolution] Update modPhpThumb.php (PR #16341)


 

Thank you for your contribution! Before your pull request can be reviewed, please sign the MODX Contributor License Agreement (CLA). This is to make sure MODX has your written permission to distribute projects containing your contributions under the appropriate license.

Please make sure the CLA has been signed for GitHub user(s): @jolichter

Once you've signed, please reply with @cla-bot check to verify and update the status of this pull request.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

smg6511 commented on core/src/Revolution/modPhpThumb.php:
```diff
@@ -302,10 +302,11 @@ public function RedirectToCachedFile()
-                header($_SERVER['SERVER_PROTOCOL'] . ' 304 Not Modified');
-                exit;
-            }
+            # Fix PHP warning: Undefined array key "HTTP_IF_MODIFIED_SINCE"
+               if (isset($_SERVER['HTTP_IF_MODIFIED_SINCE']) && ($nModified == strtotime($_SERVER['HTTP_IF_MODIFIED_SINCE'])) && isset($_SERVER['SERVER_PROTOCOL'])) {
```
You can simplify this just a bit, consider:

```suggestion

            if (isset($_SERVER['HTTP_IF_MODIFIED_SINCE'], $_SERVER['SERVER_PROTOCOL']) && $nModified === strtotime($_SERVER['HTTP_IF_MODIFIED_SINCE'])) {
```

smg6511 commented on core/src/Revolution/modPhpThumb.php:
```diff
@@ -332,13 +333,25 @@ public function SendSaveAsFileHeaderIfNeeded()
-            $this->DebugMessage('SendSaveAsFileHeaderIfNeeded() sending header: Content-Disposition: ' . (@$_GET['down'] ? 'attachment' : 'inline') . '; filename="' . $downloadfilename . '"',
-                __FILE__, __LINE__);
-            header('Content-Disposition: ' . (@$_GET['down'] ? 'attachment' : 'inline') . '; filename="' . $downloadfilename . '"');
+    # Fix PHP warning: "Undefined array key sia"
+    if (isset($_GET['sia'])) {
+        $downloadfilename = phpthumb_functions::SanitizeFilename($_GET['sia']);
+    } elseif (isset($_GET['down'])) {
+        $downloadfilename = phpthumb_functions::SanitizeFilename($_GET['down']);
+    } else {
+        $downloadfilename = 'phpThumb_generated_thumbnail' . (isset($_GET['f']) ? $_GET['f'] : 'jpg');
+    } 
+   
+    # Fix PHP warning: "Undefined array key down"
+    if (isset($downloadfilename)) {
+        if (isset($_GET['down'])) {
+            $this->DebugMessage('SendSaveAsFileHeaderIfNeeded() sending header: Content-Disposition: attachment; filename="' . $downloadfilename . '"', __FILE__, __LINE__);
+            header('Content-Disposition: attachment; filename="' . $downloadfilename . '"');
+        } else {
+            $this->DebugMessage('SendSaveAsFileHeaderIfNeeded() sending header: Content-Disposition: inline; filename="' . $downloadfilename . '"', __FILE__, __LINE__);
+            header('Content-Disposition: inline; filename="' . $downloadfilename . '"');
         }
```
Consider this alternative, which keeps the code more concise (like it originally was) but improves readability like you were going for:
```suggestion
        $inlineFilename = $_GET['sia'] ?? false;
        $attachmentFilename = $_GET['down'] ?? false;
        if ($inlineFilename || $attachmentFilename) {
            $filename = $inlineFilename ? $inlineFilename : $attachmentFilename ;
            $filename = phpthumb_functions::SanitizeFilename($filename);
        } else {
            $filename = 'phpThumb_generated_thumbnail' . ($_GET['f'] ?? 'jpg');
        }
        if ($filename) {
            $disposition = $attachmentFilename ? 'attachment' : 'inline' ;
            $this->DebugMessage('SendSaveAsFileHeaderIfNeeded() sending header: Content-Disposition: ' . $disposition . '; filename="' . $filename . '"', __FILE__, __LINE__);
            header('Content-Disposition: ' . $disposition . '; filename="' . $filename . '"');
        }
```

smg6511 reviewed (commented):
Ok, so the issue stems from the fact that, beginning with php 8.0, certain errors can not be suppressed with the @ prefix. So, this PR does the right thing by replacing the suppression with `isset` tests, and I can confirm it does the job. I'm throwing a couple suggestions for conciseness and clarity, but they aren't blockers to approval. Please respond one way or another and I'll provide approval afterward.

Mark-H commented:
Thanks for making your first contribution @jolichter! :) 

Solid suggestions from @smg6511 that I'd like to second, and fwiw those code quality issues flagged by phpcs (at least the ones in red) will also need to be addressed before this can be merged. 

We're looking to release 3.0.3 next week, would be great to get this in, but it could use some minor tweaks. Do you think you'd be able to implement those in the coming days @jolichter or would you need some help with that?

jolichter committed:
Update modPhpThumb.php

1) Fix PHP warning: Undefined array key "HTTP_IF_MODIFIED_SINCE"  (line 305)
2) Fix PHP warning: "Undefined array key sia" and "Undefined array key down" (line 335)
Thanks for the suggestion @smg6511

jolichter commented:
Thanks @Mark-H and @smg6511 !

I've updated to MODX 3.0.3-pl and changed the better suggestion of smg6511 in my 'modPhpThumb.php'. The suggestion works without problems and no error like 'HTTP_IF_MODIFIED_SINCE',  'sia' or 'down' appears. I hope I implemented this correctly and it will be included in the upcoming update? If not, I probably need your support.

PS: I would have it simple mark as resolve? Sorry  @smg6511
PS2: I resolved it (it is new for me to participate in such a big project)

JoshuaLuckers commented:
Once the PR is merged the issue will be marked as solved! 

Thanks for taking time to help make MODX better!

modxcommunity commented:
This pull request has been mentioned on **MODX Community**. There might be relevant details there:

https://community.modx.com/t/3-0-3-change-document-to-collection-no-content-anymore/6403/7
